### ISSUE TITLE
Angular router leaks components and their DOM Elements

### ISSUE THREAD

**Issue by arobinson** [angular/angular#59238]
### Which @angular/* package(s) are the source of the bug?

router

### Is this a regression?

No

### Description

Every time we navigate to a component and route away asynchronously, the Angular router is leaking the component.
This is causing severe memory issues in production code.

Steps to reproduce:
1. `git clone git@github.com:arobinson/angular-router-leak.git`
2. `pnpm install`
3. `ng serve`
4. Open http://localhost:4200/homepage in Google Chrome (or Microsoft Edge)
5. Open the chrome developer tools
6. Open the "Memory" tab
7. Take a heap snapshot
8. Filter for "andrew", verify nothing is found
9. Close the chrome developer tools
10. Click "Launch Standalone"
11. Click "Close" (it will navigate after a 2 second delay)
12. Repeat steps 10 & 11 a few times
13. Reopen the chrome developer tools
14. Open the "Memory" tab
15. Take a heap snapshot
16. Filter for "andrew", notice that the DOM element has been orphaned and will not be Garbage Collected. 

For larger components this ends up leaking over 50MB of RAM per navigation in our application.

![Image](https://github.com/user-attachments/assets/088b0baf-d795-410b-8766-efd829680f94)

Now see that somehow the problem is different with a synchronous navigation:
1. Reload the browser to clear the memory
2. Open dev tools
3. Click "Launch Standalone" again
4. This time click "Close Synchronously"
5. Repeat steps 3 & 4 a few times
15. Take a heap snapshot
16. Filter for "andrew", notice that only the last DOM element that was routed to is detached. So there is still a memory issue, but thankfully it does not grow every time we route.
![Image](https://github.com/user-attachments/assets/586b16a5-2339-438e-8d32-9977eb49d55f)

### Please provide a link to a minimal reproduction of the bug

https://github.com/arobinson/angular-router-leak

### Please provide the exception or error you saw

```true
Detached / Leaked DOM elements and their components
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 18.2.1
Node: 22.11.0
Package Manager: pnpm 9.9.0
OS: darwin arm64

Angular: 18.2.2
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1900.3
@angular-devkit/build-angular   18.2.1
@angular-devkit/core            19.0.3
@angular-devkit/schematics      19.0.3
@angular/cli                    18.2.1
@schematics/angular             19.0.3
rxjs                            7.8.1
typescript                      5.5.2
zone.js                         0.14.10
```

**Comment by JeanMeche** [comment]
If you run your app with `enableProdMode()` your can investigate more easily your leaks. It will disable the apis used by the devtools.

**Comment by arobinson** [comment]
the reproduction app has prod mode enabled, the problem still exists (https://github.com/arobinson/angular-router-leak/blob/afb18848da9aee7748670dc9b2035773dc21bf15/src/main.ts#L8).
At this point I don't know of a way we can diagnose potential memory leaks in our application as opening devtools is triggering this issue.

**Comment by JeanMeche** [comment]
Ok, I can indeed reproduce the leak. It doesn't look like the DOM elements are being retained by Angular though. 

I'm wondering if this isn't a Chrome bug itself. There is a recent ticket that looks closely related : https://issues.chromium.org/issues/376777343

**Comment by arobinson** [comment]
It looks like calling `enableProdMode` doesn't work. As you can see from the screen shot, `ngDevMode` is truthy even when `enableProdMode()` is called first.

`main.ts`:
```
import { enableProdMode } from '@angular/core';
import { platformBrowser } from '@angular/platform-browser';
import { isLocalEnvironment } from 'nw-shared';
import { from, of } from 'rxjs';
import { fromFetch } from 'rxjs/fetch';
import { catchError, finalize, switchMap, tap } from 'rxjs/operators';

import { AppModule } from './app/app.module';
import { setupNwDebugZone } from './app/shared/services/nw-ng-tools/nw-ng-tools-zone';
import { environment } from './environments/environment';
enableProdMode();
```

Using the debugger, I can confirm that the imports are creating and initializing `ngDevMode`. Therefore, but the time that the code in `main.ts` is run, it is too late

![Image](https://github.com/user-attachments/assets/45629923-6a00-4dea-aa49-84ade7a031df)

**Comment by arobinson** [comment]
Even doing this:
```
import { enableProdMode } from '@angular/core';
enableProdMode();


import { platformBrowser } from '@angular/platform-browser';
import { isLocalEnvironment } from 'nw-shared';
import { from, of } from 'rxjs';
import { fromFetch } from 'rxjs/fetch';
import { catchError, finalize, switchMap, tap } from 'rxjs/operators';

import { AppModule } from './app/app.module';
import { setupNwDebugZone } from './app/shared/services/nw-ng-tools/nw-ng-tools-zone';
import { environment } from './environments/environment';
```

does not work, just importing `enableProdMode` causes Angular to go into dev mode. Probably some global code that is executing as part of an esbuild chunk file?

**Comment by JeanMeche** [comment]
Yeah, some pre-runtime code are still under debug until the false is removed. 

You can really force it by going in the `angular.json` and changing the dev target : 

```
            "development": {
              "buildTarget": "demo:build:production"
            },
```

And run `NG_BUILD_MANGLE=0 ng serve ` to keep the symbols from being mangled.

**Comment by Delagen** [comment]
@JeanMeche I'm not sure that it's chrome bug because the same in Firefox

![Image](https://github.com/user-attachments/assets/03bb8efa-1293-470b-9dea-6e14b883f63c)

**Comment by arobinson** [comment]
This is still reproducible in Chrome Beta. According to https://issues.chromium.org/issues/376777343#comment24 it should be fixed. I can also still reproduce the issue in Chrome Canary. Perhaps the fix did not address this particular use case?

I posted [this comment](https://issues.chromium.org/issues/376777343#comment34) on the thread.

**Comment by JeanMeche** [comment]
I cannot reproduce the issue anymore on M133.

**Comment by JeanMeche** [comment]
Either way I'm closing it. It is not a Angular issue.
